### PR TITLE
fix direct assembly with equation_bcs

### PR DESCRIPTION
**Pull request by ksagiyam**
This PR fixes some equation_bc related bugs: see https://github.com/firedrakeproject/firedrake/issues/1725.

### CODE REVIEW

File: firedrake/bcs.py [hunk 1]
```diff
@@ -630,3 +630,23 @@ def homogenize(bc):
...
+    from firedrake.solving import _extract_bcs
+    bcs = _extract_bcs(bcs)
+    if form_type == 'F':
+        return tuple(bc if isinstance(bc, DirichletBC) else bc._F for bc in bcs)
+    elif form_type == 'J':
+        return tuple(bc if isinstance(bc, DirichletBC) else bc._J for bc in bcs)
+    elif form_type == 'Jp':
+        return tuple(bc if isinstance(bc, DirichletBC) else bc._Jp for bc in bcs)
+    else:
+        raise TypeError("Unknown form_type: 'form_type' must be 'F', 'J', or 'Jp'.")
```
Comment (wence-) [inline firedrake/bcs.py:652]: Given you're using strings, just do:

```
if form_type not in {"F", "J", "Jp"}:
    raise ValueError(...)
else:
   return tuple(bc if isinstance(bc, DirichletBC) else getattr(bc, f"_{form_type}") for bc in bcs)
```

File: tests/equation_bcs/test_equation_bcs_assemble.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    v = TestFunction(V)
+    a = - inner(grad(u), grad(v)) * dx
+    a1 = inner(u, v) * ds(1)
+    L1 = inner(Constant(0), v) * ds(1)
+    u = Function(V)
+    bc = EquationBC(a1 == L1, u, 1)
+
+    assemble(a, bcs=extract_equation_bc_forms(bc, 'J'))
+    assemble(a, bcs=extract_equation_bc_forms([bc], 'Jp'))
+    with pytest.raises(RuntimeError) as excinfo:
```
Comment (wence-) [inline tests/equation_bcs/test_equation_bcs_assemble.py:32]: Change the raise to `ValueError` (it's not a `TypeError`) and then assert that you get a `ValueError` here.

File: tests/equation_bcs/test_equation_bcs_assemble.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    mesh = UnitSquareMesh(2, 2)
+    V = FunctionSpace(mesh, "CG", 1)
+    u = Function(V)
+    v = TestFunction(V)
+    F = - inner(grad(u), grad(v)) * dx
+    F1 = inner(u, v) * ds(1)
+    bc = EquationBC(F1 == 0, u, 1)
+
+    assemble(F, bcs=extract_equation_bc_forms(bc, 'F'))
+    assemble(F, bcs=bc)
```
Comment (wence-) [inline tests/equation_bcs/test_equation_bcs_assemble.py:16]: Can we check for something more than "the code runs"?

File: tests/equation_bcs/test_equation_bcs_assemble.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    u = TrialFunction(V)
+    v = TestFunction(V)
+    a = - inner(grad(u), grad(v)) * dx
+    a1 = inner(u, v) * ds(1)
+    L1 = inner(Constant(0), v) * ds(1)
+    u = Function(V)
+    bc = EquationBC(a1 == L1, u, 1)
+
+    assemble(a, bcs=extract_equation_bc_forms(bc, 'J'))
+    assemble(a, bcs=extract_equation_bc_forms([bc], 'Jp'))
```
Comment (wence-) [inline tests/equation_bcs/test_equation_bcs_assemble.py:31]: Likewise here, can we check that we got the right values?

File: firedrake/assemble.py [hunk 5]
```diff
@@ -238,15 +244,18 @@ def _assemble(f, tensor=None, bcs=None, form_compiler_parameters=None,
...
         if bcs is not None:
+            if any(isinstance(bc, EquationBC) for bc in bcs):
+                if rank == 1:
+                    bcs = extract_equation_bc_forms(bcs, 'F')
+                elif rank == 2:
+                    raise RuntimeError("Unable to infer 'form_type' to be used ('J' or 'Jp'): \
+Preprocess `bcs` either with: \
+`bcs = extract_equation_bc_forms(bcs, 'J')`, \
+or with \
+`bcs = extract_equation_bc_forms(bcs, 'Jp')`.")
```
Comment (wence-) [inline firedrake/assemble.py:255]: Prefer to paste strings like this:
```python
raise RuntimeError("Part of a long string: "
                                 "Another part ")
```

Also, this message will be very long.

File: firedrake/bcs.py [hunk 6]
```diff
@@ -630,3 +630,23 @@ def homogenize(bc):
         return DirichletBC(bc.function_space(), 0, bc.sub_domain)
     else:
         raise TypeError("homogenize only takes a DirichletBC or a list/tuple of DirichletBCs")
+
+
+def extract_equation_bc_forms(bcs, form_type):
```
Comment (wence-) [inline firedrake/bcs.py:635]: Make this a method on `EquationBC`? then just call it extract forms?

File: tests/equation_bcs/test_equation_bcs_assemble.py (hunk 3 above)
Comment (ksagiyam) [reply tests/equation_bcs/test_equation_bcs_assemble.py]: Now checks assembled values.

File: tests/equation_bcs/test_equation_bcs_assemble.py (hunk 4 above)
Comment (ksagiyam) [reply tests/equation_bcs/test_equation_bcs_assemble.py]: Here, too.

**Comment by wence-** [comment]
I have rebased on top of master (fixing the merge conflicts) please have a check.

**Comment by ksagiyam** [comment]
Thanks!!
Attempted to clarify things slightly.

**Comment by wence-** [comment]
thanks.

**Pull request merged**